libcontainer/factory_linux.go

```diff
@@ -261,7 +258,28 @@ func loadState(root string) (*State, error) {
 }
 
 func validateID(id string) error {
```

Comment:
Perhaps would be good to (despite it not being exported) add a GoDoc to describe accepted formats, and outline some of it.


Looking at this PR made me go down a bit of a trip in history; let me post below, but I should probably post elsewhere

---------

I noticed the underscore being added, which was not explicitly included in the
regexp, so I wondered if that was _accidentally_ included in the past (as it's
easy to forget it's part of `\w`). I also was curious if this ID was meant to
match other ID formats (e.g., used in Moby), some of which are required to start
with an alpha-numeric character, and `._-+` only allowed as separator.

So, I went on a short trip down history;

- Validation was added in https://github.com/opencontainers/runc/commit/66e6806fd23f6c793f9478374278ed221dea3347 / https://github.com/docker-archive/libcontainer/commit/8fcda56d79d0c3652d0eb4f0317c951cfddc5b18 (https://github.com/docker-archive/libcontainer/pull/249),
  which allowed for "word" (`\w`) characters and had the underscore added explicitly. Maximum length was limited to 1024 characters (no reason mentioned for this).
- The length constraint was moved from the regex to a separate check in https://github.com/docker-archive/libcontainer/commit/de57f78590cefdf08249118e2d710908cc81b16a
- In https://github.com/opencontainers/runc/commit/d95050084baa71eb46bfeb3844cc3b806497f49c (https://github.com/opencontainers/runc/pull/31), hyphens (`-`) were added to the list of allowed characters;
  > A directory with a hyphen currently generates an InvalidId error because
  > of the regex in libcontainer. I don't believe there is any reason a
  > hyphen should be disallowed.
- Periods (`.`) were added in https://github.com/opencontainers/runc/commit/d4be3405c7f9cd88cfa4ea8cf6f6cdd4ff8ddd53 (https://github.com/opencontainers/runc/pull/647), which also removed the explicit underscore, acknowledging that it's reduntant, and already part of `\w`. The issue linked to that mentions https://github.com/opencontainers/runc/issues/635#issuecomment-196467367;
  > It is for valid filepath names because the container id needs to be written
  > to disk in something like `/run` to store state.
- Finally, `+` symbols were added in https://github.com/opencontainers/runc/commit/4629512d895e5754b93556e5093f5f0b82e5bda5 (https://github.com/opencontainers/runc/pull/675). The motivation there was https://github.com/opencontainers/runc/pull/675#issuecomment-199924556
  > I added just the + because I have a use case where the foo+abc is used as a
  > dir name and I think we should allow the containers to be named following the
  > same rules as file/dir names.
  > Yes, I agree with you that we should allow more chars that fall into this category.


So, from the above;

- The original intent was for the ID to be something that can be used as part of a path, and limiting it to "portable" characters.
- **Up until https://github.com/opencontainers/runc/pull/675, this meant IDs followed the [POSIX Portable Filename Character Set](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_282)**
 
Some what unfortunately, the runtime-spec does _NOT_ define a format (or restrictions) in any way, which means that currently runc is more restrictive than the OCI runtime spec (overall this is fine). I found multiple attempts to have this included;

- There was a discussion in https://github.com/opencontainers/runtime-spec/pull/87#discussion_r35828046 (and [on the mailing list](https://groups.google.com/a/opencontainers.org/g/dev/c/P9gZBYhiqDE)) about the ID format being under-specified, and I see a patch was created for clarification (but not merged); https://github.com/wking/opencontainer-runtime-spec/commit/2ec34c62b019431c0edebead50b56ebafe70a67c (posted at the end of this comment).
- Following https://github.com/opencontainers/runc/pull/675, @mikebrow opened a ticket in the runc repository to gather interest on formalizing the validation https://github.com/opencontainers/runc/issues/676, which did not get participation, so was closed.
- https://github.com/opencontainers/runc/pull/675#issuecomment-199915890 also mentions that the runtime spec removed the requirement for IDs to be usable as element in paths. 

To summarize my thoughts;

- It's probably good to document this function, and to outline the accepted characters (possibly outline (some of) the motivations for the chosen set of characters.
- I (personally) would be in favor of formalizing accepted characters, and would've been in favor of adopting the [POSIX Portable Filename Character Set](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_282), but the `+` addition was a bit unfortunate for that.
- Perhaps it's worth giving it another attempt to update the OCI spec (at least with a recommendation if that's acceptable)


<details>
<summary>proposed PATCH for formalizing the format from 2015</summary>

```patch
From 2ec34c62b019431c0edebead50b56ebafe70a67c Mon Sep 17 00:00:00 2001
From: "W. Trevor King" <wking@tremily.us>
Date: Mon, 10 Aug 2015 12:07:20 -0700
Subject: [PATCH] runtime: Document container ID charset and uniqueness domain

Allow the runtime to use it's own scheme, but let the caller use UUIDs
if they want.  Jonathan asked for clarification as part of #87, but
didn't suggest a particular approach [1].  When we discussed it in the
2015-08-26 meeting [2], the consensus was to just allow everything.
With container IDs like 'a/b/c' leading to state entries like
'/var/oci/containers/a/b/c/state.json'.  But that could get ugly with
container IDs that contain '../' etc.  And perhaps there are some
filesystems out there that cannot represent non-ASCII characters
(actually, I'm not even sure what charset our JSON is in ;).  I'd
rather pick this minimal charset which can handle UUIDs, and make life
easy for runtime implementers and safe for bundle consumers at a
slight cost of flexibility for bundle-authors.

There was some confusion on the list about what "ASCII letters" meant
[3], so I've explicitly listed the allowed character ranges.  Here's a
Python 3 script that shows the associated Unicode logic:

  import unicodedata

  # http://www.unicode.org/reports/tr44/tr44-4.html#GC_Values_Table
  category = {
    'Ll': 'lowercase letter',
    'Lu': 'uppercase letter',
    'Nd': 'decimal number',
    'Pd': 'dash punctuation',
  }

  for i in range(1<<7):
      char = chr(i)
      abbr = unicodedata.category(char)
      if abbr[0] in ['L', 'N'] or abbr == 'Pd':
          cat = category[abbr]
          print('{:02x} {} {}'.format(i, char, cat))

[1]: https://github.com/opencontainers/specs/pull/87#discussion_r35828046
[2]: https://github.com/opencontainers/specs/wiki/Meeting-Minutes-2015-08-26
[3]: https://groups.google.com/a/opencontainers.org/d/msg/dev/P9gZBYhiqDE/-ptpOcQ5FwAJ
     Message-Id: <7ec9cff6-c1a6-4beb-82de-16eb412bf2f8@opencontainers.org>

Reported-by: Jonathan Boulle <jonathanboulle@gmail.com>
Signed-off-by: W. Trevor King <wking@tremily.us>
---
 runtime.md | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/runtime.md b/runtime.md
index be2045b58..c9da6af9a 100644
--- a/runtime.md
+++ b/runtime.md
@@ -11,6 +11,9 @@ By providing a default location that container state is stored external applicat

 * **version** (string) Version of the OCI specification used when creating the container.
 * **id** (string) ID is the container's ID.
+  Only ASCII letters, numbers, and hyphens are valid (a–z, A–Z, 0–9, and ‘-’).
+  This value must be unique for a given host, but need not be universally unique.
+  Runtimes must allow the caller to set this ID, so that callers may choose, for example, to use [UUIDs][uuid] for universal uniqueness.
 * **pid** (int) Pid is the ID of the main process within the container.
 * **root** (string) Root is the path to the container's bundle directory.

@@ -94,3 +97,5 @@ If a hook returns a non-zero exit code, then an error is logged and the remainin

 `path` is required for a hook.
 `args` and `env` are optional.
+
+[uuid]: https://tools.ietf.org/html/rfc4122
```

</details>

Reply:
> * Finally, + symbols were added in 4629512 (#675). 

Most interestingly, as pointed out by @mrunalp [below](https://github.com/opencontainers/runc/pull/3460#discussion_r1028587711), this commit also implicitly adds comma (`,`) to the list of allowed characters, since `+-\.` is now a range from `+` (ASCII 43) to `.` (ASCII 45) and that includes ASCII 44 (`,`).

Now we need to decide whether to keep supporting `,` or not. I vote to drop `,` and deal with the consequences.